Raise repair

kylealexanderhackney opened the pull request:
## Link to ticket

https://trello.com/c/fHHnSJXg/280-as-a-user-i-need-to-be-able-to-raise-a-works-order-so-that-contractors-know-what-to-fix

## Describe this PR

### *What is the problem we're trying to solve*

A new endpoint for raising repairs. post a hact complaient raiserepair schema and data will be stored in the repairs db

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

pixeltrix reviewed (commented):
Do we want to do the renaming from V1 to V2 in this PR or leave it for a different PR?

josh-dadswell-hackney commented on RepairsApi/V1/Factories/DBModelFactory.cs:
```diff
@@ -0,0 +1,149 @@
+
+        public static Quantity ToDb(this Generated.Quantity raiseRepair)
+        {
+            if (raiseRepair.Amount.Count != 1) throw new NotSupportedException("Multiple amounts is not supported");
```
Probably needs specific test

kylealexanderhackney commented:
> Do we want to do the renaming from V1 to V2 in this PR or leave it for a different PR?

We decided to do it after this PR because the change involves renaming a folder which would cause all the files to be marked as moved. This with the some of the larger changes could make git think we deleted and added resulting in a loss of what changes happened

LBHBomotoso commented on HactGenerator/SchemaGeneration/SchemaGeneration/Program.cs:
```diff
@@ -0,0 +1,75 @@
+﻿using Newtonsoft.Json;
```
What does this class do?

josh-dadswell-hackney replied:
This is a seperate project that generates the ViewModels (/V1/Generated/https://github.com/LBHackney-IT/repairs-api-dotnet/pull/11.cs) based on the HACT spec.

LBHBomotoso commented on RepairsApi/V1/Infrastructure/AlertRegardingLocation.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace RepairsApi.V1.Infrastructure
+{
+    public class AlertRegardingLocation
```
Could this be Address instead of Location?

kylealexanderhackney replied:
These properties are defined as part of the hact specification. we dont really have much say in the names

LBHBomotoso replied:
Ok, I see

LBHBomotoso commented on RepairsApi/V1/Infrastructure/RateScheduleItem.cs:
```diff
@@ -1,20 +1,16 @@
-
-        [Column("custom_name")] public string CustomName { get; set; }
+        [Key] public Guid Id { get; set; }
+        public M3NHFScheduleOfRatesCodeC0 M3NHFSORCode { get; set; }
```
I've called this ScheduleOfRatesCodes in my migration.

josh-dadswell-hackney replied:
Is that migration in flight or already merged?

kylealexanderhackney replied:
Our migrations are done based on the naming from the hact specification

LBHBomotoso replied:
I'm still working on it. I think M3NHF is an actual product also.

josh-dadswell-hackney replied:
M3NHF is the standard of codes that can be put in through this route all other codes (including ours ones) will come through the CustomCode route

LBHBomotoso commented on RepairsApi/V1/Generated/RaiseRepair-M3SoR-v7.cs:
```diff
@@ -69249,7 +69249,7 @@ public partial class RaiseRepair
         public string DescriptionOfWork { get; set; }
```
Are these generated files used?

kylealexanderhackney replied:
Yes this file is the request object for the raise repair endpoint

josh-dadswell-hackney replied:
Yes this file is used by the controllers to receive the data and verify it against HACT required fields etc.

LBHBomotoso replied:
ok cool

LBHBomotoso commented on RepairsApi.Tests/V1/Gateways/RepairGatewayTests.cs:
```diff
@@ -35,7 +30,7 @@ public async Task Run()
 
             // assert
-            InMemoryDb.Instance.WorkOrders.Should().ContainSingle().Which.IsSameOrEqualTo(expected.ToDb());
+            InMemoryDb.Instance.WorkOrders.Should().ContainSingle().Which.IsSameOrEqualTo(expected);
```
All tests are running via in memory db? 
We should try to create a postgres image to use at some opportunity.

kylealexanderhackney replied:
running in memory for unit tests is typically preferred as it means docker is not required to run the tests and they can be done inside the IDE. the integration tests can be run against a postgres database. 
In memeory also has the benefit of not enforcing foreign key constraints so you dont have to mock large amounts of unrelated data just to run a test.

LBHBomotoso commented on RepairsApi.Tests/MockApiGateway.cs:
```diff
@@ -188,7 +188,8 @@ internal static void AddTenantInformation(string tenantReference, string propRef
                 TenancyAgreementReference = tenantReference,
                 PropertyReference = propRef,
-                TenureType = $"{code}: Rnadom descirption"
+                TenureType = $"{code}: Rnadom descirption",
```
Small typo
